### PR TITLE
[GSoC 2020] Time format updated.

### PR DESCRIPTION
**Pull request by Syed-Osama-Hussain**
Made small changes in base.py in standalone module and added helper function in main.cpp

### CODE REVIEW

File: brian2/devices/cpp_standalone/device.py [hunk 1]
```diff
@@ -1400,20 +1401,32 @@ def network_run(self, net, duration, report=None, report_period=10*second,
...
 
+        # #Code for time format function
+        # format_func_code = '''
+        #  '''
+
         # Code for a progress reporting function
         standard_code = '''
         void report_progress(const double elapsed, const double completed, const double start, const double duration)
         {
+            std::string start_format = format_time((int)start);
```
Comment (mstimberg) [inline brian2/devices/cpp_standalone/device.py]: The start time of the simulation is the "biological time", and should therefore not be formatted in the same way as the elapsed/estimated "wall time". It will often be in seconds or even milliseconds.

File: .eggs/README.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+This directory contains eggs that were downloaded by setuptools to build, test, and run plug-ins.
```
Comment (mstimberg) [inline .eggs/README.txt]: Please do not include the `.eggs` directory in the repository.

File: brian2/devices/cpp_standalone/device.py [hunk 3]
```diff
@@ -176,6 +176,7 @@ def __init__(self):
         self.net_synapses = [] 
         self.static_array_specs =[]
         self.report_func = ''
+        # self.format_func = ''
```
Comment (mstimberg) [inline brian2/devices/cpp_standalone/device.py]: Please remove unused commented code.

File: brian2/devices/cpp_standalone/device.py [hunk 4]
```diff
@@ -1400,20 +1401,32 @@ def network_run(self, net, duration, report=None, report_period=10*second,
                 for codeobj in obj._code_objects:
                     code_objects.append((obj.clock, codeobj))
 
+        # #Code for time format function
+        # format_func_code = '''
+        #  '''
+
```
Comment (mstimberg) [inline brian2/devices/cpp_standalone/device.py]: unused code

File: brian2/devices/cpp_standalone/device.py [hunk 5]
```diff
@@ -1445,6 +1458,7 @@ def network_run(self, net, duration, report=None, report_period=10*second,
                                           'each run has to use the same (or '
                                           'none).')
             self.report_func = report_func
+            # self.format_func = format_func_code
```
Comment (mstimberg) [inline brian2/devices/cpp_standalone/device.py]: more unused code

File: brian2/devices/cpp_standalone/templates/network.cpp [hunk 6]
```diff
@@ -122,7 +122,8 @@ void Network::run(const double duration, void (*report_func)(const double, const
         _last_run_completed_fraction = 1.0;
     }
     if (report_func)
-    {
+    {   
+        // cout<<duration<<"1266666666666666666666666"
```
Comment (mstimberg) [inline brian2/devices/cpp_standalone/templates/network.cpp]: more unused code

File: brian2/devices/cpp_standalone/templates/objects.cpp [hunk 7]
```diff
@@ -324,4 +324,4 @@ void _dealloc_arrays();
 #endif
 
 
-{% endmacro %}
+{% endmacro %}
```
Comment (mstimberg) [inline brian2/devices/cpp_standalone/templates/objects.cpp]: Please do not remove the newline at the end of the file.

File: brian2/devices/cpp_standalone/templates/main.cpp [hunk 8]
```diff
@@ -20,6 +20,37 @@
...
+
+	ss << minutes;
+	std::string mins = ss.str();
+	
+	ss << time_in_s;
+	std::string secs = ss.str();
+
+	if(hours < 1){
+		if(minutes < 1){
+			return secs + " s"
```
Comment (mstimberg) [inline brian2/devices/cpp_standalone/templates/main.cpp]: Sorry, didn't see it before: there's a `;` missing at the end of the line here, making compilation fail.

**Comment by Syed-Osama-Hussain** [comment]
@mstimberg Sorry for the extra files (I thought gitignore will handle it). Can I update the .gitignore file so this doesn't happen again? Also, this was not my final commit. I wanted to check if this was working on the CI tool, since I didn't have visual studio at that time. Thanks for pointing out the issues!

File: .eggs/README.txt (hunk 2 above)
Comment (Syed-Osama-Hussain) [reply .eggs/README.txt]: Noted.

File: brian2/devices/cpp_standalone/device.py (hunk 1 above)
Comment (Syed-Osama-Hussain) [reply brian2/devices/cpp_standalone/device.py]: Noted.

**Comment by mstimberg** [comment]
Sorry for the premature feedback... Yes, these cases should ideally be handled by `.gitignore`, feel free to add at least `output` to it. It doesn't hurt to have `.egg` in there as well, even though I don't see how it got created. Did you run `python setup.py bdist_egg`? Either way, please add a short comment when you are ready for feedback.

Also note that you can activate your repository on https://travis-ci.org/ , it can then run tests for your branches before you open a pull request.

**Comment by Syed-Osama-Hussain** [comment]
I ran the installation command:
`pip install -e .`
and ran one of the test files (maybe this caused the files to be created). I'll get back to you as soon as I complete the task. Thanks for the CI suggestion.

**Comment by Syed-Osama-Hussain** [comment]
Hello @mstimberg. I have updated the time format. I was having some problems with my internet due to which I couldn't install VS build tools. Sorry for the delay.
I will commit again after I've attempted the bonus tasks. In the mean time, I would like to request you for your feedback on this commit.

**Comment by Syed-Osama-Hussain** [comment]
Attempted to solve #1162 and  #1173. Hello @mstimberg. I've declared a global object in main.cpp for estimation of time. Would you approve of this approach or should I find a better way to use the estimator object?

File: brian2/core/network.py [hunk 9]
```diff
@@ -52,6 +52,7 @@
 
 
 def _format_time(time_in_s):
+    print(time_in_s,"hereeeeeee")
```
Comment (mstimberg) [inline brian2/core/network.py:55]: Quite sure that shouldn't be here anymore :smile:

File: brian2/devices/cpp_standalone/device.py [hunk 10]
```diff
@@ -750,7 +751,7 @@ def generate_main_source(self, writer):
                                                            main_lines=main_lines,
                                                            code_lines=self.code_lines,
                                                            code_objects=list(self.code_objects.values()),
-                                                           report_func=self.report_func,
+                                                           report_func=self.report_func,time_format_func=self.time_format_func,
```
Comment (mstimberg) [inline brian2/devices/cpp_standalone/device.py:754]: Please move the second assignment into a separate line.

File: brian2/devices/cpp_standalone/device.py [hunk 11]
```diff
@@ -1400,20 +1401,58 @@ def network_run(self, net, duration, report=None, report_period=10*second,
...
+            }else{
+                return hrs + "h " + mins + "m " + secs + "s";
+            }
+        }'''
+
         # Code for a progress reporting function
         standard_code = '''
         void report_progress(const double elapsed, const double completed, const double start, const double duration)
         {
+            std::string duration_format = format_time((int)duration);
```
Comment (mstimberg) [inline brian2/devices/cpp_standalone/device.py:1440]: `duration_format` is not used anywhere, so it can be deleted.

File: brian2/devices/cpp_standalone/templates/main.cpp [hunk 12]
```diff
@@ -20,6 +21,13 @@
 
 #include <iostream>
 #include <fstream>
+#include <string>
+#include <sstream>
+#include <vector>
+
+Estimator estimator;
```
Comment (mstimberg) [inline brian2/devices/cpp_standalone/templates/main.cpp:28]: I think a better alternative to the global variable here would be to put it into the report function with static scope. I guess just adding
```C++
static Estimator estimator;
```
to the beginning of the function would work. This would have the advantage that there's no `estimator` object when we don't use the report function in the first place.

File: brian2/devices/cpp_standalone/templates/network.cpp [hunk 13]
```diff
@@ -122,7 +122,7 @@ void Network::run(const double duration, void (*report_func)(const double, const
         _last_run_completed_fraction = 1.0;
     }
     if (report_func)
-    {
+    {
```
Comment (mstimberg) [inline brian2/devices/cpp_standalone/templates/network.cpp:125]: some unnecessary whitespace here

File: brian2/core/network.py (hunk 9 above)
Comment (Syed-Osama-Hussain) [reply brian2/core/network.py:55]: Yes. Really sorry for this mistake.

File: brian2/devices/cpp_standalone/device.py (hunk 11 above)
Comment (Syed-Osama-Hussain) [reply brian2/devices/cpp_standalone/device.py:1440]: I've used duration_format at line 1446 to format the duration.

File: brian2/devices/cpp_standalone/templates/main.cpp (hunk 12 above)
Comment (Syed-Osama-Hussain) [reply brian2/devices/cpp_standalone/templates/main.cpp:28]: Noted. Thanks

File: brian2/devices/cpp_standalone/device.py (hunk 11 above)
Comment (Syed-Osama-Hussain) [reply brian2/devices/cpp_standalone/device.py:1440]: Should the duration be formatted?

**Comment by Syed-Osama-Hussain** [comment]
Sorry for all the extra commits.

File: brian2/devices/cpp_standalone/device.py (hunk 11 above)
Comment (mstimberg) [reply brian2/devices/cpp_standalone/device.py]: No, `duration` (and `completed*duration`) refer to "biological time" and are in seconds or even milliseconds and should not be formatted this way. Only the  `elapsed` and the `remaining` time are "wall clock time" and are seconds/minutes/hours/...

Comment (Syed-Osama-Hussain) [reply brian2/devices/cpp_standalone/device.py]: Okay.

**Comment by mstimberg** [comment]
Apologies for the long delay in reviewing. I did not want to give further comments on PRs made as part of GSoC applications after the application deadline, to not give participants the false impression that this work was still taken into account for the application. I am going to close all "GSoC pull request" now, but of course we'd be more than happy to see you as a contributor in the future. Thanks again for your contribution!
